packages/block-editor/src/components/provider/index.js

```diff
@@ -68,31 +68,47 @@ class BlockEditorProvider extends Component {
 		const {
 			getBlocks,
 			isLastBlockChangePersistent,
+			getLastBlockAttributesChange,
 			__unstableIsLastBlockChangeIgnored,
 		} = registry.select( 'core/block-editor' );
 
 		let blocks = getBlocks();
 		let isPersistent = isLastBlockChangePersistent();
+		let lastBlockAttributesChange = getLastBlockAttributesChange();
 
 		this.unsubscribe = registry.subscribe( () => {
 			const {
 				onChange,
 				onInput,
+				onBlockAttributesChange,
 			} = this.props;
+
 			const newBlocks = getBlocks();
 			const newIsPersistent = isLastBlockChangePersistent();
+			const newLastBlockAttributesChange = getLastBlockAttributesChange();
+
 			if (
 				newBlocks !== blocks && (
 					this.isSyncingIncomingValue ||
 					__unstableIsLastBlockChangeIgnored()
 				)
 			) {
-				this.isSyncingIncomingValue = false;
+				this.isSyncingIncomingValue = null;
 				blocks = newBlocks;
 				isPersistent = newIsPersistent;
+				lastBlockAttributesChange = newLastBlockAttributesChange;
 				return;
 			}
 
+			if ( newLastBlockAttributesChange !== lastBlockAttributesChange ) {
```

Comment:
If we did something like this, I think it would have to be a tuple of the entire set of attributes changes, as otherwise I'm not sure when we would ever "clean up" old attributes changes? And more work to iterate through the changes.

Something like:

```
[ lastAttributesChanges, newAttributesChanges ]
```

... where each has the shape:

```
{ [ clientId ]: { [ attribute ]: change } }
```

Reply:
Another option: In @youknowriad's #16075, he [implemented this reducer](https://github.com/WordPress/gutenberg/pull/16075/files#diff-ff2edcc560dbdf06e4a49898f00e323aR986) to always reset to `null` when the action dispatched is anything other than a block attributes update. That way we could infer that a change has occurred if the value resulting from `getLastBlockAttributesChange` is anything other than `null`.

The only thing [unclear to me](https://github.com/WordPress/gutenberg/pull/16075/files#r299628880) is whether we can safely assume that the value will be non-`null` as long as we need it to be.

Reply:
> Another option: In @youknowriad's #16075, he [implemented this reducer](https://github.com/WordPress/gutenberg/pull/16075/files#diff-ff2edcc560dbdf06e4a49898f00e323aR986) to always reset to `null` when the action dispatched is anything other than a block attributes update. That way we could infer that a change has occurred if the value resulting from `getLastBlockAttributesChange` is anything other than `null`.

I implemented this in 085c9eb64. It seems to work quite well, and gains all of the benefits you were hoping from your suggestion.

Reply:
>as otherwise I'm not sure when we would ever "clean up" old attributes changes?

In hindsight, we would have had this problem with any data structure. It arises from the fact that you want to "cleanup" after checking for a change and reacting on it. This can't be done easily if the data is shared like in my suggestion. You would need to keep local flags everywhere you use it.

>I implemented this in 085c9eb. It seems to work quite well, and gains all of the benefits you were hoping from your suggestion as far as simplifying the implementation to avoid onBlockAttributesUpdate and instead fetch the updates directly within resetEditorBlocks.

This is perfect for what we need now. Much cleaner 👍